---
title: Transcriptomic analysis of high-oil transgenic cowpea
subtitle: Exploration of gene expression changes
short_title: High-oil transgenic cowpea # Delete if not required

author:  Jose Barrero
affiliation: CSIRO Agriculture and Food # Or group/team
photo: resources/img/Jose_photo.jpg

output: DSreport::project_summary
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  results = 'asis',
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  fig.align = 'center'
)

###Load libraries
library(tidyverse)
library(kableExtra)
library(png)
library(ggplot2)
library(grid)

###reading files

read_csv('data/embryo_all_normalised.csv')
embryo_expression <- read_csv('data/embryo_all_normalised.csv')

read_csv('data/cowpea_geneatlas.csv')
gene_atlas <- read_csv('data/cowpea_geneatlas.csv')

read_csv('data/sample_oilcontent.csv')
oilcontent <- read_csv('data/sample_oilcontent.csv')


###merging gene files

gene_atlas <- gene_atlas <- rename(gene_atlas, Id=1)
colnames(gene_atlas)
gene_annotation <- select(gene_atlas, 1,2, 25)
view(full_join(embryo_expression, gene_annotation, by = c('Id')))
embryo_expression_annotation <- full_join(embryo_expression, gene_annotation, by = c('Id'))
embryo_expression_annotation <- rename(embryo_expression_annotation, annotation=11, gene_name=10)

####tidying expression table

tidy_embryo_expression <- embryo_expression_annotation %>% 
  gather(key = Sample_name, value = ammount, -Id, -annotation, -gene_name)
tidy_embryo_expression <- tidy_embryo_expression %>% 
  mutate(sample = Sample_name)
tidy_embryo_expression <- separate(tidy_embryo_expression, Sample_name, into = c('sample_name', 'tissue', 'stage'), sep = "_")
tidy_embryo_expression <- separate(tidy_embryo_expression, sample_name, into = c('todelete', 'genotype'), sep = "m.")
tidy_embryo_expression <- rename(tidy_embryo_expression, gene_id=1, expression=8)
tidy_embryo_expression <- tidy_embryo_expression %>% select(-todelete)

write_csv(tidy_embryo_expression, 'results/tidy_embryo_expression.csv')

####making sample description plot

Oilcontent_plot <- ggplot(data = oilcontent,
  mapping = aes(x = stage, y = oil_contentDW, colour = genotype, size = weight_mg)
) + annotation_custom(rasterGrob(img, 
                               width = unit(1,"npc"),
                               height = unit(1,"npc")), 
                    -Inf, Inf, -Inf, Inf) + geom_point() 

Oilcontent_plot <- Oilcontent_plot +  
  labs(title = "Oil content in transgenic and wildtype cowpeas",
       #caption = 'ADD CAPTION',
       x = "Seed developmental stage",
       y = "Oil content (% DW)",
       size = "Seed weight (mg)",
       colour = "Genotype") + 
  #scale_x_continuous())+
  scale_y_continuous(limits=c(0, 10))+
  
  theme(
    panel.grid.minor = element_blank(),
    plot.title = element_text(face = "bold"),
    strip.background = element_blank(),
    panel.grid.major = element_line(size = 1),
    axis.title = element_text(size = 10, face = "bold"),
    )

ggsave(filename = "results/oilcontent_plot.png", plot = Oilcontent_plot,
       width = 15, height = 12, dpi = 300, units = "cm")

### #Making heatmap

expn_data <- read_csv("results/tidy_embryo_expression.csv")

# Data processing
# Extremely basic, not ideal for a proper analysis
# 1: Remove genes with 0 expression
# 2: Convert to log scale
# 3: Normalise each gene to mean=0, sd=1

expn_scaled <- expn_data %>% 
	group_by(gene_name) %>% 
	filter(sum(expression) > 100) %>% 
  #filter(var(expression) > 50)
	mutate(
		log_expn = log2(expression + 1),
		norm_expn = (log_expn - mean(log_expn))/sd(log_expn)
	) %>% 
	ungroup()

# Check a plot
# Looks pretty good, but alphabetical order for genes not helpful

expn_scaled %>% 
	ggplot(aes(y = gene_name, x = sample, fill = norm_expn)) +
	geom_tile() +
	# Too many genes to show the names or the axis ticks so remove them
	scale_y_discrete(labels = NULL, breaks = NULL) +
	# Convert the Brewer RdBu palette into a continuous colour scale
	scale_fill_distiller(palette = "RdBu")

# Need expression in wide format for clustering
# The final step converts it from a tibble into a data.frame with
# the gene names as the 'rownames' of the data.frame
# This removes the gene names from the data (so we just have numeric data) 
# and makes them more metadata for each row (which will be important later)

expn_table <- expn_scaled %>%
	select(gene_name, norm_expn, sample) %>% 
	spread(key = sample, value = norm_expn) %>% 
	column_to_rownames("gene_name") 
	
# Can't use hierarchical clustering on this many genes because you need to
# make all-to-all pairwise comparisons (this is probably why the heatmap
# function didn't work). For this sort of preliminary data, I like to use k-means
# clustering with an arbitrary choice of the number of clusters (k) to get a quick
# look at what's in the data

expn_clustered <- cluster::clara(expn_table, k = 10)



# The cluster result (expn_clustered) has an element called clustering that 
# shows what cluster each gene is assigned to. We can access this with expn_clustered$clustering
# We sort this so that the genes from each cluster are together, then get 
# the gene names with names() -- the rownames from the expn_table data.frame
# are used to give names to the clustering output.

gene_order <- names(sort(expn_clustered$clustering))

# Now redo the plot with one change, the gene names are first converted to a
# factor, with the levels of the factor being our gene order.
# This makes the genes be plotted in the correct order by ggplot.

heatmap_plot <- expn_scaled %>% 
	mutate(gene_name = factor(gene_name, levels = gene_order)) %>% 
	ggplot(aes(y = gene_name, x = sample, fill = norm_expn)) +
	geom_tile() +
  facet_grid(~ genotype, switch = "x", scales = "free_x", space = "free_x") +
	scale_y_discrete(labels = NULL, breaks = NULL) +
	scale_fill_distiller(palette = "RdBu") +
  scale_x_discrete(labels=c("norm.T_Embryo_1" = "1", 
                            "norm.T_Embryo_2" = "2", 
                            "norm.T_Embryo_3" = "3", 
                            "norm.T_Embryo_4" = "4", 
                            "norm.WT_Embryo_1" = "1",
                            "norm.WT_Embryo_2" = "2",
                            "norm.WT_Embryo_3" = "3",
                            "norm.WT_Embryo_4" = "4"))


heatmap_plot <- heatmap_plot +  
  labs(title = "Gene expression heatmap during seed developmet",
       subtitle = "Comparison between the wildtype and the high-oil transgenic",
       caption = '(Raw RNAseq data analysed by J. Verdier)',
       x = "Sample developmental stage",
       y = "Genes",
       fill = "Normalized expression") +
  
  theme(
    plot.title = element_text(face = "bold"),
    #strip.background = element_blank(),
    axis.title = element_text(size = 10, face = "bold"),
    #axis.text.x = element_text(angle=45),
    legend.position="bottom")
  

ggsave(filename = "results/heatmap_plot.png", plot = heatmap_plot,
       width = 15, height = 20, dpi = 300, units = "cm")




```


# Introduction
My name is *Jose Barrero* and I am a Research Scientist in Agriculture and Food. I worked mainly with cereal crops but I recently started leading the Cowpea Team in Black Mountain. I have not done any coding before but I generate large datasets that need complex analysis, so that is why I decided to learn a bit of R.

# My Project
My project is about exploring gene changes in a **transgenic high-oil cowpea** line we have made. Our transgenic line (named 15B) has more oil in the seeds that the wildtype line (named IT86). We collected RNA samples from developing seeds from 15B and IT86 at 4 different stages and performed a trancriptomic RNAseq analysis. The goal of this project is to evaluate the gene expression changes bewteen the two geneotypes and identify genes related with the production of high oil. The plan making was making a heatmap illustrating the genes changes across the samples and identify clusters of genes with similar expression patterns.


## Description of the samples


We first analysed the oil content in seeds from wildtype and transgenic seeds at develomental stages (Figure 1). The transgenic line showed a significant increased in oil content from stage 3, and remains high until stage 5 (matured dry seed). 
We performed RNAseq in stages 1 to 4 in both gentypes. 

**Oil content analysis in our cowpea samples**
```{r standard-plot, out.width='60%', fig.align='center', fig.height= 4, fig.width=6, fig.cap="Yet another gapminder plot"}
Oilcontent_plot <- ggplot(data = oilcontent,
  mapping = aes(x = stage, y = oil_contentDW, colour = genotype, size = weight_mg)
) + annotation_custom(rasterGrob(img, 
                               width = unit(1,"npc"),
                               height = unit(1,"npc")), 
                    -Inf, Inf, -Inf, Inf) + geom_point() 

Oilcontent_plot <- Oilcontent_plot +  
  labs(title = "Oil content in transgenic and wildtype cowpeas",
       #caption = 'ADD CAPTION',
       x = "Seed developmental stage",
       y = "Oil content (% DW)",
       size = "Seed weight (mg)",
       colour = "Genotype") + 
  #scale_x_continuous())+
  scale_y_continuous(limits=c(0, 10))+
  
  theme(
    panel.grid.minor = element_blank(),
    plot.title = element_text(face = "bold"),
    strip.background = element_blank(),
    panel.grid.major = element_line(size = 1),
    axis.title = element_text(size = 10, face = "bold"),
    )
```

Your figure and table captions are automatically numbered and can be referenced in the text
if needed: see eg. Table \@ref(tab:mytable) and Figure \@ref(fig:standard-plot)




To get tables formatting correctly, use `knitr::kable` to convert the table to html format. If
you also want to have alternate row highlighting, pass the result to `kable_styling('striped')` 
from the `kableExtra` package.

**Tables**
```{r mytable, out.width='100%', echo = T}
knitr::kable(head(gapminder, n = 5), format = "html", caption = "A table of data") %>% 
  kable_styling("striped")
```

**Images from a file**

![](resources/img/tidyverse.png){width=100px}

**Plots from R**
```{r standard-plot, out.width='60%', fig.align='center', fig.height= 4, fig.width=6, fig.cap="Yet another gapminder plot"}
gapminder %>% 
  filter(year == 1987) %>% 
  ggplot(aes(x = gdpPercap, y = lifeExp, colour = continent, size = pop)) +
  geom_point() +
  scale_x_log10(labels = function(b) format(b, scientific = F)) +
  scale_size(trans = "sqrt", guide = "none") +
  scale_color_brewer(palette = "Set1") +
  theme_linedraw() +
  theme(legend.position = "bottom")
```

Your figure and table captions are automatically numbered and can be referenced in the text
if needed: see eg. Table \@ref(tab:mytable) and Figure \@ref(fig:standard-plot)

# My Digital Toolbox

What digital tools have you been using in your project? Do you expect that everything will be able 
to be completed within R, or will you need to work with multiple tools to get the right result?
Which of the digital skills needed for your project have you learned since starting Data School?

You can use all the usual R markdown features in writing a project summary, including lists:

* R - dplyr, ggplot, ...
* Python
* SQL

## Favourite tool (optional)

Is there a tool/package/function in particular that you've enjoyed using? Give it a special shout out here. What about this tool makes it your favourite?

![](https://raw.githubusercontent.com/tidyverse/ggplot2/master/man/figures/logo.png){.pull-right width=100px}

No prizes for guessing mine:

# My time went ...

What parts of your project take the most time and effort? Were there any surprising challenges you
encountered, and how did you solve them?

# Next steps

What further steps do you wish your project could take? Or are there any new digital skills that you
are keen to develop as a result of your involvement in the Data School?

# My Data School Experience

This summary is mostly about your project. However we would also like to hear about other
parts of your Data School experience. What aspects of the program did you really enjoy? Have you
tried applying the skills you have learned in your daily work? Have you been able to transfer this 
knowledge to your team members? Any descriptions of the personal impact the program has 
had are welcome here as well!
